tools/shared/src/main/scala/org/scalajs/core/tools/javascript/JSDesugaring.scala

```diff
@@ -1566,13 +1619,28 @@ object JSDesugaring {
           val newExpr = transformExpr(expr)
 
           if (semantics.asInstanceOfs == Unchecked) {
+            @inline def boolOr0(tree: js.Tree): js.Tree =
+              if (isStrongMode) js.BinaryOp(JSBinaryOp.||, tree, js.IntLiteral(0))
+              else tree
+
             (charCode: @switch) match {
               case 'Z'             => !(!newExpr)
-              case 'B' | 'S' | 'I' =>
-                js.BinaryOp(JSBinaryOp.|, newExpr, js.IntLiteral(0))
-              case 'J'             => genCallHelper("uJ", newExpr)
-              case 'F'             => genFround(newExpr)
-              case 'D'             => js.UnaryOp(JSUnaryOp.+, newExpr)
+              case 'B' | 'S' | 'I' => or0(boolOr0(newExpr))
+              case 'J' =>
+                if (isStrongMode)
+                  js.BinaryOp(JSBinaryOp.||, newExpr, js.VarRef(js.Ident("$longZero")))
+                else
+                  genCallHelper("uJ", newExpr)
+              case 'F' =>
+                if (isStrongMode)
+                  genCallHelper("uF", newExpr)
+                else
+                  genFround(boolOr0(newExpr))
```

Comment:
`boolOr0` doesn't make sense here, since we are not in strong mode, therefore it is just the tree itself.